src/tools/kata-ctl/src/iptables.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2023 Alec Pemberton, Juanaiga Okugas
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+//use clap::{App, arg, Parser, SubCommand, Command};
+use reqwest::{Url};
+use std::{fs};
+use anyhow::{Result, Context};//Context
+use shim_interface::shim_mgmt::client::MgmtClient;
+use crate::args::{IptablesCommand, IpTablesArguments};//Commands
+use thiserror::Error;
+use std::time::Duration;
+//use super::*;
+
+//kata-proxy management API endpoint, without code would not know the location of the unix sockets
+const DEFAULT_TIMEOUT: u64 = 30;
+const IP_TABLES_SOCKET: &str = "unix:///run/vc/sbs/{sandbox_id}/ip_tables";
+const IP6_TABLES_SOCKET: &str = "unix:///run/vc/sbs/{sandbox_id}/ip6_tables";
+
+#[derive(Error, Debug)]
+pub enum Error{
+    #[error("Invalid Container ID {0}")]
+    InvalidContainerID(String),
+}
+
+//Verify Id for validating sandboxID
+pub fn verify_id(id:&str) -> Result<(), Error>{
+    let mut chars = id.chars();
+
+    let valid = matches!(chars.next(), Some(first) if first.is_alphanumeric()
+    &&id.len() >1
+    && chars.all(|c| c.is_alphanumeric() || ['.', '-', '_'].contains(&c)));
+
+    match valid {
+        true => Ok(()),
+        false => Err(Error::InvalidContainerID(id.to_string())),
+    }
+}
+
+pub async fn handle_iptables(args: IptablesCommand) -> Result<(), anyhow::Error> {//pub fn handle
+    //checking for subcommand entered form user 
+    match args.subcommand() {//.subcommand()
+        IpTablesArguments::Get{sandbox_id, v6} =>{//Some(("get", get_matches)) => {
+            // retrieve the sandbox ID from the command line arguments
+            let sandbox_id = sandbox_id;//get_matches.value_of("sandbox-id")?;
+
+            let is_ipv6 = v6;//v6;get_matches.is_present("v6");
+           
+            verify_id(sandbox_id)?;
+            // generate the appropriate URL for the iptables request to connect Kata to agent within guest
+            let url = if *is_ipv6 {
+                Url::parse(&format!("{}{}", IP6_TABLES_SOCKET, sandbox_id))?
+            } else {
+                Url::parse(&format!("{}{}", IP_TABLES_SOCKET, sandbox_id))?
```

Comment:
I think you'll find that the expanded string passed to `Url::parse()` isn't what you expect.

I'd recommend writing a function that returns either the ipv4 or ipv6 API URL instead. The function will only be a few lines long and would look like this:

```rust
fn mk_ip_tables_socket_path(sandbox_id: &str, ipv6: bool) -> String {
    // Fill me in ;)
}
```

Once you have that function, you can add a few unit tests to ensure that whatever arguments you specify what comes out looks like a reasonable URL/URI.